codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/parse/JsonParserGenerator.kt

```diff
@@ -510,23 +514,30 @@ class JsonParserGenerator(
                                 """,
                                 *codegenScope,
                             )
-                            withBlock("variant = match key.to_unescaped()?.as_ref() {", "};") {
+                            withBlock("match key.to_unescaped()?.as_ref() {", "};") {
                                 for (member in shape.members()) {
                                     val variantName = symbolProvider.toMemberName(member)
                                     rustBlock("${jsonName(member).dq()} =>") {
                                         if (member.isTargetUnit()) {
                                             rustTemplate(
                                                 """
                                                 #{skip_value}(tokens)?;
-                                                Some(#{Union}::$variantName)
+                                                variant = Some(#{Union}::$variantName)
                                                 """,
                                                 "Union" to returnSymbolToParse.symbol, *codegenScope,
                                             )
                                         } else {
-                                            withBlock("Some(#T::$variantName(", "))", returnSymbolToParse.symbol) {
-                                                deserializeMember(member)
+                                            rust("let (value, is_not_null) =")
+                                            deserializeMember(member)
+                                            rust(".map_or((None, false), |v| (Some(v), true));")
```

Comment:
Something seems off here. If we got into this variant match arm, then we have an explicitly set variant, right? So it shouldn't matter if it's null or not?

Reply:
In fact, I think a variant being explicitly set to `null` is an error condition that we could probably short-circuit on.

Reply:
> then we have an explicitly set variant, right? So it shouldn't matter if it's null or not?

This was because of `unwrap_or_default`

> I think a variant being explicitly set to null is an error condition that we could probably short-circuit on

Nice! Changed to reflect this